internal/epoch/epoch_advance.go

```diff
@@ -36,6 +37,10 @@ func shouldAdvance(bms []blob.Metadata, minEpochDuration time.Duration, countThr
 		totalSize += bm.Length
 	}
 
+	if totalSize >= totalSizeBytesThresholdHard {
+		return true
+	}
+
```

Comment:
@Lyndon-Li Please remove this check from this PR (and the corresponding tests). At the moment, we are not going to change the conditions for compaction. Also, if it were to do it, then it should be separately in its own PR.

```suggestion
```

Reply:
If we don't add this change, I am afraid the epoch maintenance in the quick maintenance doesn't make any difference --- since the `minEpochDuration` is 24 hour, even though quick maintenance launches the epoch maintenance every 1 hour, it actually does nothing since the epoch is not advanced.

Reply:
Running epoch maintenance in quick maintenance would make a difference. Specifically regarding epoch advancement, it would advance it within an hour of the reaching the minimum epoch duration age. Otherwise, in the worst case scenario the epoch would be advance close to 24 hours after being reaching the minimum epoch age.

Reply:
Yes, but that is not enough, or the epoch advancing is not timely enough once a burst of index generation happens. The consequence is that the next index compaction takes huge memory.

Reply:
You can lower the minimum epoch duration for the repository.

Reply:
That doesn't work well for the case I mentioned:
1. We can not predict that the coming backups generate large number of indexes. Or in another word, this doesn't happen for every backup
2. This is not timely enough. E.g., say we set the duration to 4 hour, but a backup may complete in 1 hour and generates large number of indexes

Reply:
That doesn't work well for the case I mentioned:
1. We can not predict that the coming backups generate large number of indexes. Or in another word, this doesn't happen for every backup
2. This is not timely enough. E.g., say we set the duration to 4 hour, but a backup may complete in 1 hour and generates large number of indexes

Reply:
May I know the concern here? are we expecting the epochs to always be regular in terms of time?

Reply:
@julio-lopez 

> Is it safe?
Sorry, I didn't see this comment.

Below is my thinking:

- The index compaction won't delete the uncompacted/superseded indexes immediately. After they are compacted in a quick maintenance, they will be deleted in the following FULL maintenance ONLY when they are old enough (controlled by `CleanupSafetyMargin`, by default 4 hours).
- Therefore, adding the hard limit ostensibly increase the risk of data lose, but risk is actually decided by `CleanupSafetyMargin`. If data safety is the concern, users should increase `CleanupSafetyMargin`.
- Additionally, at present, we support a customized duration time as shorter as 10min. So users should have already been using this experience --- advance the epoch in a short time.

Data safety is always with top priority, but scalability and performance is also important. I am not sure if the changes in this PR is 100% correct, but it brings a possible solution of considering the two aspects together. So appreciate your further comments.

Reply:
@Lyndon-Li 

You raise an important point about the safety of deleting compacted indexes. Indeed a safety margin is needed so clients that are in the process of reading an old index snapshot can read that snapshot.

The proposed change adds a new criterion for advancing the epoch (`totalSize >= totalSizeBytesThresholdHard`). The original protocol design uses 2 other criteria to reach consensus about:(a) when it is time to advance and epoch; and (b) when it safe to assume that the epoch will no longer be modified, so it can be compacted.

The new criterion appears to be violating one of the original assumptions. It _may_ be that the proposed change is safe, but it is not yet clear, at least not to me.

Reply:
OK, I got you points and understand your concern of data safety.

And it is also a fact that the current behavior of epoch advancing brings burst burden to repo maintenance (epoch index compaction), as index creations are never linearly streamed along with time, but influx with backups.
Let's see if this brings problems to other cases, then we can come back to discuss this.